File: bindings/browser/examples/mux-roundtrip.ts
// ASPIRATIONAL EXAMPLE — no such binding package exists yet.
//
// This file shows the target ergonomics for a future browser/WASM binding of
// Mediaway's container mux/demux capability. The browser host goes through
// `wasm-bindgen` (never the C ABI) — see `docs/spec/c-ffi.md` Tier C and
// `../README.md`. Written as if `@mediaway/browser` already existed on npm.
//
// Mirrors `examples/mux_roundtrip.rs`: build a fragmented MP4 with one H.264
// video track and one AAC audio track, then demux the same bytes back.

import init, { Muxer, Demuxer } from "@mediaway/browser";
import type { Packet } from "@mediaway/browser";

async function main(): Promise<void> {
  // The WASM module must be instantiated (fetched + compiled) once before any
  // other call. Every call after this resolves is synchronous-feeling.
  await init();

  const videoTimeBase = { num: 1, den: 30 }; // 30 fps
  const audioTimeBase = { num: 1, den: 48_000 }; // 48 kHz
  const frameCount = 90; // 3 s at 30 fps

  // 1. Create a muxer — starts "open": tracks can still be registered.
  const muxer = new Muxer();

  const videoTrackId = muxer.addTrack({
    kind: "video",
    codec: "h264",
    timeBase: videoTimeBase,
    width: 1920,
    height: 1080,
    extraData: new Uint8Array(0),
  });

  const audioTrackId = muxer.addTrack({
    kind: "audio",
    codec: "aac",
    timeBase: audioTimeBase,
    extraData: new Uint8Array(0),
    sampleRate: 48_000,
    channels: 2,
  });

  // 2. begin() closes track registration and switches the muxer to "live".
  muxer.begin();

  const videoPayload = new Uint8Array([0x00, 0x00, 0x00, 0x01]); // placeholder NAL
  const audioPayload = new Uint8Array([0xff, 0xf1]); // placeholder AAC frame

  for (let frame = 0; frame < frameCount; frame++) {
    muxer.pushPacket({
      streamId: videoTrackId,
      pts: frame,
      dts: frame,
      duration: 1,
      isKeyframe: frame % 30 === 0,
      isDiscard: false,
      payload: videoPayload,
    });

    muxer.pushPacket({
      streamId: audioTrackId,
      pts: frame * 1_600,
      dts: frame * 1_600,
      duration: 1_600,
      isKeyframe: true,
      isDiscard: false,
      payload: audioPayload,
    });
  }

  muxer.flush();

  // 3. Pull out the muxed bytes — the muxer never touches storage/network
  // itself; the caller decides where they go (upload, MediaSource, disk, ...).
  const mp4Bytes = muxer.pollBytes();
  muxer.free(); // release the WASM-side muxer; JS GC can't reach into WASM memory

  console.log(`mux-roundtrip: ${frameCount} frames -> ${mp4Bytes.byteLength} bytes of fMP4`);

  // 4. Demux the same bytes back, streaming them in as if freshly received.
  const demuxer = new Demuxer();
  demuxer.pushBytes(mp4Bytes);

  console.log(`mux-roundtrip: demuxer sees ${demuxer.streams.length} stream(s)`);
  for (const stream of demuxer.streams) {
    if (stream.kind === "video") {
      console.log(`  stream ${stream.id} - ${stream.codec} ${stream.width}x${stream.height}`);
    } else {
      console.log(`  stream ${stream.id} - ${stream.codec}`);
    }
  }

  let videoPackets = 0;
  let audioPackets = 0;
  let packet: Packet | undefined;
  while ((packet = demuxer.pollPacket()) !== undefined) {
    if (packet.streamId === videoTrackId) {
      videoPackets++;
    } else {
      audioPackets++;
    }
  }
  demuxer.free();

  console.log(`mux-roundtrip: recovered ${videoPackets} video + ${audioPackets} audio packets`);
}

main().catch((err: unknown) => {
  console.error(err);
});

File: bindings/csharp/examples/MuxRoundtrip.cs
// MuxRoundtrip.cs — Mediaway C# quick start.
//
// The `Mediaway.Common`/`Mediaway.Container` packages this targets are real —
// see bindings/csharp/src/ and docs/adr/0017-csharp-binding-package-layout.md.
// This file mirrors examples/mux_roundtrip.rs and matches the real, shipped
// API: IDisposable wrappers over native handles, exceptions instead of raw
// error codes, PascalCase members.
//
// Revision note: PollBytes()/PollPacket()/Streams return IMemoryOwner<byte>-
// backed types (`Packet`, `StreamDescriptor` are themselves IDisposable) so
// the native buffer is released deterministically on Dispose, never via a
// finalizer — see Mediaway.Common.Interop.NativeOwnedMemoryManager's own
// comment on why a finalizer there would risk a use-after-free (CA2015)
// instead of just a late free.
//
// Scenario: mux one H.264 video track + one AAC audio track into a
// fragmented MP4 (sans-io — the muxer/demuxer never touch files or sockets;
// the caller always owns the bytes), then demux the result back and count
// the recovered packets.

using System;
using System.Buffers;
using Mediaway.Common;
using Mediaway.Container;

var frameRate = new Rational(1, 30);
var audioTimeBase = new Rational(1, 48_000);
const int frameCount = 90; // 3 s at 30 fps

// ── 1. Register tracks (Open state) ─────────────────────────────────────────
using var muxer = new Muxer();

uint videoTrackId = muxer.AddTrack(new VideoTrackInfo
{
    Id = 0,
    Codec = CodecKind.H264,
    TimeBase = frameRate,
    Width = 1920,
    Height = 1080,
    ExtraData = Array.Empty<byte>(),
});

uint audioTrackId = muxer.AddTrack(new AudioTrackInfo
{
    Id = 1,
    Codec = CodecKind.Aac,
    TimeBase = audioTimeBase,
    ExtraData = Array.Empty<byte>(),
    SampleRate = 48_000,
    Channels = 2,
});

// ── 2. Begin() closes track registration and hands back the live session ───
// (AddTrack / PushPacket / Flush throw MediawayException — translated from
// the native C ABI error codes — on failure; omitted here for the happy path.)
using MuxerSession session = muxer.Begin();

byte[] fakeNalUnit = { 0x00, 0x00, 0x00, 0x01 };
byte[] fakeAdtsFrame = { 0xFF, 0xF1 };

for (int i = 0; i < frameCount; i++)
{
    session.PushPacket(new Packet
    {
        StreamId = videoTrackId,
        Pts = i,
        Dts = i,
        Duration = 1,
        IsKeyframe = i % 30 == 0,
        IsDiscard = false,
        Payload = fakeNalUnit,
    });

    session.PushPacket(new Packet
    {
        StreamId = audioTrackId,
        Pts = i * 1_600L,
        Dts = i * 1_600L,
        Duration = 1_600,
        IsKeyframe = true,
        IsDiscard = false,
        Payload = fakeAdtsFrame,
    });
}

session.Flush();

// ── 3. Pull the muxed bytes — the muxer never writes to disk itself ────────
// Zero-Copy over the native buffer; PollBytes() returns an IMemoryOwner<byte>
// (not a byte[]) precisely so this Dispose is where the native free happens —
// never via a finalizer (see this file's header revision note).
using IMemoryOwner<byte> mp4Bytes = session.PollBytes();
Console.WriteLine($"Muxed {frameCount} frames into {mp4Bytes.Memory.Length} bytes of fMP4.");

// ── 4. Demux the same bytes back ────────────────────────────────────────────
using var demuxer = new Demuxer();
demuxer.PushBytes(mp4Bytes.Memory.Span); // PushBytes can also be called incrementally as bytes arrive.

var streams = demuxer.Streams;
Console.WriteLine($"Demuxer discovered {streams.Count} stream(s):");
foreach (StreamDescriptor stream in streams)
{
    using (stream) // StreamDescriptor owns its ExtraData buffer — dispose after reading it.
    {
        string shape = stream.Geometry is { } geometry
            ? $"{geometry.Width}x{geometry.Height}"
            : "no geometry";
        Console.WriteLine($"  stream {stream.Id}: {stream.Codec} ({shape})");
    }
}

int videoPacketCount = 0;
int audioPacketCount = 0;
while (demuxer.PollPacket() is { } packet)
{
    using (packet) // Packet owns its Payload buffer when it comes from PollPacket — dispose it.
    {
        if (packet.StreamId == videoTrackId)
        {
            videoPacketCount++;
        }
        else
        {
            audioPacketCount++;
        }
    }
}

Console.WriteLine($"Recovered {videoPacketCount} video + {audioPacketCount} audio packets.");
